<a href="https://colab.research.google.com/github/pujaroy280/DATA602Assignment7/blob/main/Puja_Roy_07_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Assignment 7**

# **Weeks 8 & 9 - Pandas**
* In this homework assignment, you will explore and analyze a public dataset of your choosing. Since this assignment is “open-ended” in nature, you are free to expand upon the requirements below. However, you must meet the minimum requirments as indicated in each section.

* You must use Pandas as the **primary tool** to process your data.

* The preferred method for this analysis is in a .ipynb file. Feel free to use whichever platform of your choosing.  
 * https://www.youtube.com/watch?v=inN8seMm7UI (Getting started with Colab).

* Your data should need some "work", or be considered "dirty".  You must show your skills in data cleaning/wrangling.

### **Some data examples:**
•	https://www.data.gov/

•	https://opendata.cityofnewyork.us/

•	https://datasetsearch.research.google.com/

•	https://archive.ics.uci.edu/ml/index.php

### **Resources:**

•	https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html

•	https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html


### **Headings or comments**
**You are required to make use of comments, or headings for each section.  You must explain what your code is doing, and the results of running your code.**  Act as if you were giving this assignment to your manager - you must include clear and descriptive information for each section.

### **You may work as a group or indivdually on this assignment.**


# Introduction

In this section, please describe the dataset you are using.  Include a link to the source of this data.  You should also provide some explanation on why you choose this dataset.

For this assignment, I am planning to leverage the New York Times Best Sellers Dataset from Kaggle: https://www.kaggle.com/datasets/dhruvildave/new-york-times-best-sellers

**Description of Data:** All New York Times Best Sellers List from 2010 to 2019 (10 years)

**Dataset updated:**
Dec 11, 2020

This dataset is interesting because I love reading books that are based in New York. As a New York resident, I enjoy watching shows and reading books based on NY's setting. I want to analyze this dataset to observe which books from New York were top-ranked, the price of books, the weeks that the books were on the list for the best selling, and if the dataset contains the books that I am planning to read over the summer.

______________
# Data Exploration
Import your dataset into your .ipynb, create dataframes, and explore your data.  

Include:

* Summary statistics means, medians, quartiles,
* Missing value information
* Any other relevant information about the dataset.  



In [2]:
import pandas as pd

# URL of the CSV file
# Fetch raw data from my GitHub profile
url = "https://raw.githubusercontent.com/pujaroy280/DATA602Assignment7/main/bestsellers.csv"

# Load the data into a DataFrame
df = pd.read_csv(url)

In [3]:
# Display the first few rows of the DataFrame
print(df.head())

  published_date      list_name list_name_encoded  rank         isbn13  \
0     2010-01-03  Chapter Books     chapter-books     1  9780316036245   
1     2010-01-03  Chapter Books     chapter-books     2  9780439023481   
2     2010-01-03  Chapter Books     chapter-books     3  9780439023498   
3     2010-01-03  Chapter Books     chapter-books     4  9780763644109   
4     2010-01-03  Chapter Books     chapter-books     5  9780385738934   

       isbn10                    title  \
0  0316036242         WITCH AND WIZARD   
1  0439023483         THE HUNGER GAMES   
2  0439023491            CATCHING FIRE   
3  0763644102  THE MAGICIAN’S ELEPHANT   
4  0385738935                   FALLEN   

                                     author  \
0  James Patterson and Gabrielle Charbonnet   
1                           Suzanne Collins   
2                           Suzanne Collins   
3                            Kate DiCamillo   
4                               Lauren Kate   

                   

In [4]:
# Display columns
print(df.columns)

Index(['published_date', 'list_name', 'list_name_encoded', 'rank', 'isbn13',
       'isbn10', 'title', 'author', 'description', 'amazon_product_url',
       'price', 'weeks_on_list'],
      dtype='object')


In [5]:
# Display summary statistics
print(df.describe())

               rank         price  weeks_on_list
count  61430.000000  61430.000000   61430.000000
mean       3.000000      3.863068      24.706267
std        1.414225      8.789769      63.534314
min        1.000000      0.000000       0.000000
25%        2.000000      0.000000       0.000000
50%        3.000000      0.000000       2.000000
75%        4.000000      0.000000      16.000000
max        5.000000    150.000000     607.000000


In [6]:
# 2. Missing value information
missing_values = df.isnull().sum()
print("\nMissing Value Information:")
print(missing_values)


Missing Value Information:
published_date           0
list_name                0
list_name_encoded        0
rank                     0
isbn13                   4
isbn10                1284
title                    0
author                  69
description           8119
amazon_product_url       0
price                    0
weeks_on_list            0
dtype: int64


In [7]:
# 3. Other relevant information
print("\nNumber of unique authors:", df['author'].nunique())
print("Number of unique titles:", df['title'].nunique())
print("Number of unique list names:", df['list_name'].nunique())


Number of unique authors: 3609
Number of unique titles: 6577
Number of unique list names: 59


# Data Wrangling
Create a subset of your original data and perform the following.  

1. Modify multiple column names.

2. Look at the structure of your data – are any variables improperly coded? Such as strings or characters? Convert to correct structure if needed.

3. Fix missing and invalid values in data.

4. Create new columns based on existing columns or calculations.

5. Drop column(s) from your dataset.

6. Drop a row(s) from your dataset.

7. Sort your data based on multiple variables.

8. Filter your data based on some condition.

9. Convert all the string values to upper or lower cases in one column.

10. Check whether numeric values are present in a given column of your dataframe.

11. Group your dataset by one column, and get the mean, min, and max values by group.
  * Groupby()
  * agg() or .apply()

12. Group your dataset by two columns and then sort the aggregated results within the groups.

**You are free (and should) to add on to these questions.  Please clearly indicate in your assignment your answers to these questions.**

In [19]:
df.head()

published_date      list_name list_name_encoded  rank         isbn13  \
0     2010-01-03  Chapter Books     chapter-books     1  9780316036245   
1     2010-01-03  Chapter Books     chapter-books     2  9780439023481   
2     2010-01-03  Chapter Books     chapter-books     3  9780439023498   
3     2010-01-03  Chapter Books     chapter-books     4  9780763644109   
4     2010-01-03  Chapter Books     chapter-books     5  9780385738934   

       isbn10                    title  \
0  0316036242         WITCH AND WIZARD   
1  0439023483         THE HUNGER GAMES   
2  0439023491            CATCHING FIRE   
3  0763644102  THE MAGICIAN’S ELEPHANT   
4  0385738935                   FALLEN   

                                     author  \
0  James Patterson and Gabrielle Charbonnet   
1                           Suzanne Collins   
2                           Suzanne Collins   
3                            Kate DiCamillo   
4                               Lauren Kate   

                                         description  \
0  One of each, brother and sister, flex their ne...   
1  In a dystopian future, a girl fights for survi...   
2     The protagonist of "The Hunger Games" returns.   
3  An orphan in search of his sister follows a fo...   
4  Thwarted love among misfits at a Savannah, Ga....   

                                  amazon_product_url  price  weeks_on_list  
0  https://www.amazon.com/Witch-Wizard-James-Patt...  17.99              1  
1  https://www.amazon.com/The-Hunger-Games-Suzann...  17.99             67  
2  https://www.amazon.com/Catching-Fire-Hunger-Ga...  17.99             16  
3  https://www.amazon.com/The-Magicians-Elephant-...  16.99             15  
4  https://www.amazon.com/Fallen-Lauren-Kate/dp/0...  17.99              2

In [24]:
# 1. Modify multiple column names
df.rename(columns={'list_name': 'Book_Material','title': 'Book_Title', 'description': 'Summary'}, inplace=True)
df.tail()

published_date                  Book_Material  \
61425     2019-12-29  Young Adult Paperback Monthly   
61426     2019-12-29  Young Adult Paperback Monthly   
61427     2019-12-29  Young Adult Paperback Monthly   
61428     2019-12-29  Young Adult Paperback Monthly   
61429     2019-12-29  Young Adult Paperback Monthly   

                   list_name_encoded  rank         isbn13      isbn10  \
61425  young-adult-paperback-monthly     1  9780142412145  0142412147   
61426  young-adult-paperback-monthly     2  9780593109069  0593109066   
61427  young-adult-paperback-monthly     3  9781481438261  1481438263   
61428  young-adult-paperback-monthly     4  9780553496710  0553496719   
61429  young-adult-paperback-monthly     5  9780525555377  0525555374   

                     Book_Title  \
61425               LET IT SNOW   
61426        LOOKING FOR ALASKA   
61427             LONG WAY DOWN   
61428    THE SUN IS ALSO A STAR   
61429  TURTLES ALL THE WAY DOWN   

                                               author Summary  \
61425  John Green, Maureen Johnson and Lauren Myracle     NaN   
61426                                      John Green     NaN   
61427                                  Jason Reynolds     NaN   
61428                                     Nicola Yoon     NaN   
61429                                      John Green     NaN   

                                      amazon_product_url  price  weeks_on_list  
61425  https://www.amazon.com/Let-It-Snow-Holiday-Rom...    0.0              0  
61426  https://www.amazon.com/Looking-Alaska-John-Gre...    0.0              0  
61427  https://www.amazon.com/Long-Way-Down-Jason-Rey...    0.0              0  
61428  https://www.amazon.com/Sun-Also-Star-Nicola-Yo...    0.0              0  
61429  https://www.amazon.com/Turtles-All-Down-John-G...    0.0              0

In [25]:
# 2. Check data structure and correct if needed
# No corrections needed based on provided information

In [27]:
# Remove rows with missing values
df_cleaned = df.dropna()

# Alternatively, you can remove columns with missing values by specifying axis=1
# df_cleaned = df.dropna(axis=1)

# Display the shape of the cleaned DataFrame to see the effect of removing missing values
print("Shape of original DataFrame:", df.shape)
print("Shape of cleaned DataFrame:", df_cleaned.shape)

Shape of original DataFrame: (61430, 12)
Shape of cleaned DataFrame: (51997, 12)


In [29]:
df_cleaned.tail()

published_date          Book_Material      list_name_encoded  rank  \
61420     2019-12-29  Young Adult Hardcover  young-adult-hardcover     1   
61421     2019-12-29  Young Adult Hardcover  young-adult-hardcover     2   
61422     2019-12-29  Young Adult Hardcover  young-adult-hardcover     3   
61423     2019-12-29  Young Adult Hardcover  young-adult-hardcover     4   
61424     2019-12-29  Young Adult Hardcover  young-adult-hardcover     5   

              isbn13      isbn10                        Book_Title  \
61420  9781250170996  1250170990  CHILDREN OF VIRTUE AND VENGEANCE   
61421  9781524714680  1524714682                ONE OF US IS LYING   
61422  9781250170972  1250170974        CHILDREN OF BLOOD AND BONE   
61423  9781534451568  1534451560                   FIVE FEET APART   
61424  9780062498533  0062498533                   THE HATE U GIVE   

                                                  author  \
61420                                       Tomi Adeyemi   
61421                                    Karen M McManus   
61422                                       Tomi Adeyemi   
61423  Rachael Lippincott with Mikki Daughtry and Tob...   
61424                                       Angie Thomas   

                                                 Summary  \
61420  Zélie must stop the threat of civil war in Orï...   
61421  For five students, a detour into detention end...   
61422  Zélie fights to restore magic to the land of O...   
61423  Stella and Will are in love, but they can't ge...   
61424  A 16-year-old girl sees a police officer kill ...   

                                      amazon_product_url  price  weeks_on_list  
61420  https://www.amazon.com/Children-Virtue-Vengean...    0.0              2  
61421  https://www.amazon.com/One-Us-Lying-Karen-McMa...    0.0             99  
61422  https://www.amazon.com/Children-Blood-Bone-Leg...    0.0             93  
61423  https://www.amazon.com/Five-Feet-Apart-Rachael...    0.0             54  
61424  https://www.amazon.com/Hate-U-Give-Angie-Thoma...    0.0            146

In [44]:
# 4. Create new columns
# Create a new column for the total number of Book_Material
"""
In this code, we use the apply() method to apply a function to each entry in the Book_Material column.
The lambda function lambda x: len(x) calculates the length of each entry (i.e., the total number of characters) in the Book_Material column, and the result is stored in the new column Total_Book_Material.
"""
df_cleaned['Total_Book_Material'] = df_cleaned['Book_Material'].apply(lambda x: len(x))
df_cleaned.tail()

<ipython-input-44-d7ba40e908a1>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['Total_Book_Material'] = df_cleaned['Book_Material'].apply(lambda x: len(x))


published_date          Book_Material      list_name_encoded  rank  \
61420     2019-12-29  Young Adult Hardcover  young-adult-hardcover     1   
61421     2019-12-29  Young Adult Hardcover  young-adult-hardcover     2   
61422     2019-12-29  Young Adult Hardcover  young-adult-hardcover     3   
61423     2019-12-29  Young Adult Hardcover  young-adult-hardcover     4   
61424     2019-12-29  Young Adult Hardcover  young-adult-hardcover     5   

              isbn13      isbn10                        Book_Title  \
61420  9781250170996  1250170990  CHILDREN OF VIRTUE AND VENGEANCE   
61421  9781524714680  1524714682                ONE OF US IS LYING   
61422  9781250170972  1250170974        CHILDREN OF BLOOD AND BONE   
61423  9781534451568  1534451560                   FIVE FEET APART   
61424  9780062498533  0062498533                   THE HATE U GIVE   

                                                  author  \
61420                                       Tomi Adeyemi   
61421                                    Karen M McManus   
61422                                       Tomi Adeyemi   
61423  Rachael Lippincott with Mikki Daughtry and Tob...   
61424                                       Angie Thomas   

                                                 Summary  \
61420  Zélie must stop the threat of civil war in Orï...   
61421  For five students, a detour into detention end...   
61422  Zélie fights to restore magic to the land of O...   
61423  Stella and Will are in love, but they can't ge...   
61424  A 16-year-old girl sees a police officer kill ...   

                                      amazon_product_url  price  \
61420  https://www.amazon.com/Children-Virtue-Vengean...    0.0   
61421  https://www.amazon.com/One-Us-Lying-Karen-McMa...    0.0   
61422  https://www.amazon.com/Children-Blood-Bone-Leg...    0.0   
61423  https://www.amazon.com/Five-Feet-Apart-Rachael...    0.0   
61424  https://www.amazon.com/Hate-U-Give-Angie-Thoma...    0.0   

       weeks_on_list  Total_Book_Material  
61420              2                   21  
61421             99                   21  
61422             93                   21  
61423             54                   21  
61424            146                   21

In [45]:
# 5. Drop column(s) from the dataset
df_cleaned.drop(columns=['Total_Book_Material'], inplace=True)  # Dropping the newly created column for demonstration
df_cleaned.tail()

<ipython-input-45-bebec4681d23>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned.drop(columns=['Total_Book_Material'], inplace=True)  # Dropping the newly created column for demonstration


published_date          Book_Material      list_name_encoded  rank  \
61420     2019-12-29  Young Adult Hardcover  young-adult-hardcover     1   
61421     2019-12-29  Young Adult Hardcover  young-adult-hardcover     2   
61422     2019-12-29  Young Adult Hardcover  young-adult-hardcover     3   
61423     2019-12-29  Young Adult Hardcover  young-adult-hardcover     4   
61424     2019-12-29  Young Adult Hardcover  young-adult-hardcover     5   

              isbn13      isbn10                        Book_Title  \
61420  9781250170996  1250170990  CHILDREN OF VIRTUE AND VENGEANCE   
61421  9781524714680  1524714682                ONE OF US IS LYING   
61422  9781250170972  1250170974        CHILDREN OF BLOOD AND BONE   
61423  9781534451568  1534451560                   FIVE FEET APART   
61424  9780062498533  0062498533                   THE HATE U GIVE   

                                                  author  \
61420                                       Tomi Adeyemi   
61421                                    Karen M McManus   
61422                                       Tomi Adeyemi   
61423  Rachael Lippincott with Mikki Daughtry and Tob...   
61424                                       Angie Thomas   

                                                 Summary  \
61420  Zélie must stop the threat of civil war in Orï...   
61421  For five students, a detour into detention end...   
61422  Zélie fights to restore magic to the land of O...   
61423  Stella and Will are in love, but they can't ge...   
61424  A 16-year-old girl sees a police officer kill ...   

                                      amazon_product_url  price  weeks_on_list  
61420  https://www.amazon.com/Children-Virtue-Vengean...    0.0              2  
61421  https://www.amazon.com/One-Us-Lying-Karen-McMa...    0.0             99  
61422  https://www.amazon.com/Children-Blood-Bone-Leg...    0.0             93  
61423  https://www.amazon.com/Five-Feet-Apart-Rachael...    0.0             54  
61424  https://www.amazon.com/Hate-U-Give-Angie-Thoma...    0.0            146

In [48]:
# 6. Drop row(s) from the dataset
# Let's drop rows where Author is missing
df_cleaned.dropna(subset=['author'], inplace=True)
df_cleaned.head()

<ipython-input-48-ef2f65294554>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned.dropna(subset=['author'], inplace=True)


published_date  Book_Material list_name_encoded  rank         isbn13  \
0     2010-01-03  Chapter Books     chapter-books     1  9780316036245   
1     2010-01-03  Chapter Books     chapter-books     2  9780439023481   
2     2010-01-03  Chapter Books     chapter-books     3  9780439023498   
3     2010-01-03  Chapter Books     chapter-books     4  9780763644109   
4     2010-01-03  Chapter Books     chapter-books     5  9780385738934   

       isbn10               Book_Title  \
0  0316036242         WITCH AND WIZARD   
1  0439023483         THE HUNGER GAMES   
2  0439023491            CATCHING FIRE   
3  0763644102  THE MAGICIAN’S ELEPHANT   
4  0385738935                   FALLEN   

                                     author  \
0  James Patterson and Gabrielle Charbonnet   
1                           Suzanne Collins   
2                           Suzanne Collins   
3                            Kate DiCamillo   
4                               Lauren Kate   

                                             Summary  \
0  One of each, brother and sister, flex their ne...   
1  In a dystopian future, a girl fights for survi...   
2     The protagonist of "The Hunger Games" returns.   
3  An orphan in search of his sister follows a fo...   
4  Thwarted love among misfits at a Savannah, Ga....   

                                  amazon_product_url  price  weeks_on_list  
0  https://www.amazon.com/Witch-Wizard-James-Patt...  17.99              1  
1  https://www.amazon.com/The-Hunger-Games-Suzann...  17.99             67  
2  https://www.amazon.com/Catching-Fire-Hunger-Ga...  17.99             16  
3  https://www.amazon.com/The-Magicians-Elephant-...  16.99             15  
4  https://www.amazon.com/Fallen-Lauren-Kate/dp/0...  17.99              2

In [54]:
# 7. Sort the data based on multiple variables
df_cleaned_sorted = df_cleaned.sort_values(by=['price', 'weeks_on_list'], ascending=[False, True])
df_cleaned_sorted

published_date            Book_Material        list_name_encoded  rank  \
9042      2012-01-01  Hardcover Graphic Books  hardcover-graphic-books     3   
9273      2012-01-15  Hardcover Graphic Books  hardcover-graphic-books     4   
13100     2012-09-16  Hardcover Graphic Books  hardcover-graphic-books     1   
13217     2012-09-23  Hardcover Graphic Books  hardcover-graphic-books     3   
5508      2011-05-22  Hardcover Graphic Books  hardcover-graphic-books     4   
...              ...                      ...                      ...   ...   
61230     2019-12-15             Series Books             series-books     1   
61322     2019-12-22             Series Books             series-books     3   
61320     2019-12-22             Series Books             series-books     1   
61412     2019-12-29             Series Books             series-books     3   
61411     2019-12-29             Series Books             series-books     2   

              isbn13      isbn10                        Book_Title  \
9042   9781401234515  1401234518             DC COMICS: THE NEW 52   
9273   9781607065159  1607065150  THE WALKING DEAD OMNIBUS, VOL. 2   
13100  9781401234591  1401234593            THE INVISIBLES OMNIBUS   
13217  9781401234591  1401234593            THE INVISIBLES OMNIBUS   
5508   9780785146339  0785146334                     THOR: OMNIBUS   
...              ...         ...                               ...   
61230  9781419739033  1419739034              DIARY OF A WIMPY KID   
61322  9780545791427  0545791421                      HARRY POTTER   
61320  9781419739033  1419739034              DIARY OF A WIMPY KID   
61412  9780545791427  0545791421                      HARRY POTTER   
61411  9781419739033  1419739034              DIARY OF A WIMPY KID   

                          author  \
9042                     various   
9273    Robert Kirman and others   
13100  Grant Morrison and others   
13217  Grant Morrison and others   
5508             Walter Simonson   
...                          ...   
61230                Jeff Kinney   
61322                 JK Rowling   
61320                Jeff Kinney   
61412                 JK Rowling   
61411                Jeff Kinney   

                                                 Summary  \
9042   In September, DC reintroduced all its characte...   
9273   This deluxe editions comprises issues 24 throu...   
13100  This massive tome contains the entire run of G...   
13217  This massive tome contains the entire run of G...   
5508   Just in time for the film, or a nice change of...   
...                                                  ...   
61230        The travails and challenges of adolescence.   
61322  A wizard hones his conjuring skills in the ser...   
61320        The travails and challenges of adolescence.   
61412  A wizard hones his conjuring skills in the ser...   
61411        The travails and challenges of adolescence.   

                                      amazon_product_url  price  weeks_on_list  
9042   https://www.amazon.com/DC-Comics-The-New-52/dp...  150.0              1  
9273   https://www.amazon.com/Walking-Dead-Omnibus-Vo...  150.0              1  
13100  https://www.amazon.com/The-Invisibles-Various/...  150.0              1  
13217  https://www.amazon.com/The-Invisibles-Various/...  150.0              2  
5508   https://www.amazon.com/THOR-WALTER-SIMONSON-OM...  125.0              1  
...                                                  ...    ...            ...  
61230  https://www.amazon.com/Diary-Wimpy-Kid-Hard-Lu...    0.0            559  
61322  https://www.amazon.com/Harry-Potter-And-Order-...    0.0            559  
61320  https://www.amazon.com/Diary-Wimpy-Kid-Hard-Lu...    0.0            560  
61412  https://www.amazon.com/Harry-Potter-And-Order-...    0.0            560  
61411  https://www.amazon.com/Diary-Wimpy-Kid-Hard-Lu...    0.0            561  

[51997 rows x 12 columns]

In [58]:
# 8. Filter the data based on some condition
# Filter data in the author column to observe the books that Toni Morrison wrote.
author_name = df_cleaned_sorted[df_cleaned_sorted['author'] == 'Toni Morrison']
author_name

published_date            Book_Material        list_name_encoded  rank  \
30914     2015-05-10        Hardcover Fiction        hardcover-fiction     5   
59851     2019-08-25  Trade Fiction Paperback  trade-fiction-paperback     2   
31099     2015-05-17        Hardcover Fiction        hardcover-fiction     5   
59933     2019-09-01  Trade Fiction Paperback  trade-fiction-paperback     4   

              isbn13      isbn10          Book_Title         author  \
30914  9780307594174  0307594173  GOD HELP THE CHILD  Toni Morrison   
59851  9781400033416  1400033411             BELOVED  Toni Morrison   
31099  9780307594174  0307594173  GOD HELP THE CHILD  Toni Morrison   
59933  9781400033416  1400033411             BELOVED  Toni Morrison   

                                                 Summary  \
30914  Her mother’s rejection shapes the life of a da...   
59851  Winner of the 1988 Pulitzer Prize for fiction....   
31099  Her mother’s rejection shapes the life of a da...   
59933  Winner of the 1988 Pulitzer Prize for fiction....   

                                      amazon_product_url  price  weeks_on_list  
30914  https://www.amazon.com/God-Help-Child-A-novel/...    0.0              1  
59851  https://www.amazon.com/Beloved-Toni-Morrison/d...    0.0              1  
31099  https://www.amazon.com/God-Help-Child-A-novel/...    0.0              2  
59933  https://www.amazon.com/Beloved-Toni-Morrison/d...    0.0              2

In [59]:
# 8. Filter the data based on some condition
# Filter data in the author column to observe the books that Stephen King wrote.
author_name = df_cleaned_sorted[df_cleaned_sorted['author'] == 'Stephen King']
author_name

published_date            Book_Material        list_name_encoded  rank  \
8520      2011-11-27        Hardcover Fiction        hardcover-fiction     1   
8631      2011-12-04        Hardcover Fiction        hardcover-fiction     2   
8731      2011-12-11        Hardcover Fiction        hardcover-fiction     2   
8831      2011-12-18        Hardcover Fiction        hardcover-fiction     2   
8930      2011-12-25        Hardcover Fiction        hardcover-fiction     1   
...              ...                      ...                      ...   ...   
60252     2019-09-29  Trade Fiction Paperback  trade-fiction-paperback     3   
20208     2013-12-08           E-Book Fiction           e-book-fiction     4   
60337     2019-10-06  Trade Fiction Paperback  trade-fiction-paperback     3   
60427     2019-10-13  Trade Fiction Paperback  trade-fiction-paperback     3   
60519     2019-10-20  Trade Fiction Paperback  trade-fiction-paperback     5   

              isbn13      isbn10 Book_Title        author  \
8520   9781451627282  1451627289   11/22/63  Stephen King   
8631   9781451627282  1451627289   11/22/63  Stephen King   
8731   9781451627282  1451627289   11/22/63  Stephen King   
8831   9781451627282  1451627289   11/22/63  Stephen King   
8930   9781451627282  1451627289   11/22/63  Stephen King   
...              ...         ...        ...           ...   
60252  9781982127794  1982127791         IT  Stephen King   
20208  9781451627305        None   11/22/63  Stephen King   
60337  9781982127794  1982127791         IT  Stephen King   
60427  9781982127794  1982127791         IT  Stephen King   
60519  9781982127794  1982127791         IT  Stephen King   

                                                 Summary  \
8520   An English teacher travels back to 1958 by way...   
8631   An English teacher travels back to 1958 by way...   
8731   An English teacher travels back to 1958 by way...   
8831   An English teacher travels back to 1958 by way...   
8930   An English teacher travels back to 1958 by way...   
...                                                  ...   
60252  The fears of seven teenagers are rekindled in ...   
20208  An English teacher travels back to 1958 by way...   
60337  The fears of seven teenagers are rekindled in ...   
60427  The fears of seven teenagers are rekindled in ...   
60519  The fears of seven teenagers are rekindled in ...   

                                      amazon_product_url  price  weeks_on_list  
8520   https://www.amazon.com/11-22-63-A-Novel/dp/145...   35.0              1  
8631   https://www.amazon.com/11-22-63-A-Novel/dp/145...   35.0              2  
8731   https://www.amazon.com/11-22-63-A-Novel/dp/145...   35.0              3  
8831   https://www.amazon.com/11-22-63-A-Novel/dp/145...   35.0              4  
8930   https://www.amazon.com/11-22-63-A-Novel/dp/145...   35.0              5  
...                                                  ...    ...            ...  
60252  https://www.amazon.com/Stephen-King-ebook/dp/B...    0.0             23  
20208  https://www.amazon.com/11-22-63-A-Novel/dp/145...    0.0             24  
60337  https://www.amazon.com/Stephen-King-ebook/dp/B...    0.0             24  
60427  https://www.amazon.com/Stephen-King-ebook/dp/B...    0.0             25  
60519  https://www.amazon.com/Stephen-King-ebook/dp/B...    0.0             26  

[233 rows x 12 columns]

In [61]:
# 9. Convert all the string values to upper or lower cases in one column
df_cleaned_sorted['list_name_encoded'] = df_cleaned_sorted['list_name_encoded'].str.upper()  # Converting list values to uppercase
df_cleaned_sorted.head()

published_date            Book_Material        list_name_encoded  rank  \
9042      2012-01-01  Hardcover Graphic Books  HARDCOVER-GRAPHIC-BOOKS     3   
9273      2012-01-15  Hardcover Graphic Books  HARDCOVER-GRAPHIC-BOOKS     4   
13100     2012-09-16  Hardcover Graphic Books  HARDCOVER-GRAPHIC-BOOKS     1   
13217     2012-09-23  Hardcover Graphic Books  HARDCOVER-GRAPHIC-BOOKS     3   
5508      2011-05-22  Hardcover Graphic Books  HARDCOVER-GRAPHIC-BOOKS     4   

              isbn13      isbn10                        Book_Title  \
9042   9781401234515  1401234518             DC COMICS: THE NEW 52   
9273   9781607065159  1607065150  THE WALKING DEAD OMNIBUS, VOL. 2   
13100  9781401234591  1401234593            THE INVISIBLES OMNIBUS   
13217  9781401234591  1401234593            THE INVISIBLES OMNIBUS   
5508   9780785146339  0785146334                     THOR: OMNIBUS   

                          author  \
9042                     various   
9273    Robert Kirman and others   
13100  Grant Morrison and others   
13217  Grant Morrison and others   
5508             Walter Simonson   

                                                 Summary  \
9042   In September, DC reintroduced all its characte...   
9273   This deluxe editions comprises issues 24 throu...   
13100  This massive tome contains the entire run of G...   
13217  This massive tome contains the entire run of G...   
5508   Just in time for the film, or a nice change of...   

                                      amazon_product_url  price  weeks_on_list  
9042   https://www.amazon.com/DC-Comics-The-New-52/dp...  150.0              1  
9273   https://www.amazon.com/Walking-Dead-Omnibus-Vo...  150.0              1  
13100  https://www.amazon.com/The-Invisibles-Various/...  150.0              1  
13217  https://www.amazon.com/The-Invisibles-Various/...  150.0              2  
5508   https://www.amazon.com/THOR-WALTER-SIMONSON-OM...  125.0              1

In [63]:
# 10. Check whether numeric values are present in a given column of the dataframe
print("\nNumeric values in 'Summary':", df_cleaned_sorted['Summary'].str.isnumeric().any())


Numeric values in 'Summary': False


In [66]:
# 11. Group the dataset by one column and get the mean, min, and max values by group

# Group the dataset by one column and get the mean, min, and max values by group
grouped_by_book_material = df_cleaned_sorted.groupby('Book_Material').agg({'price': ['mean', 'min', 'max']})

# Display the result
print(grouped_by_book_material)

                                          price               
                                           mean    min     max
Book_Material                                                 
Animals                                0.000000   0.00    0.00
Audio Fiction                          0.000000   0.00    0.00
Audio Nonfiction                       0.000000   0.00    0.00
Celebrities                            0.000000   0.00    0.00
Chapter Books                         18.536286   9.99   39.99
Childrens Middle Grade                 0.000000   0.00    0.00
Childrens Middle Grade Hardcover       0.000000   0.00    0.00
Combined Print Fiction                 0.000000   0.00    0.00
Combined Print Nonfiction              0.000000   0.00    0.00
Combined Print and E-Book Fiction      0.000000   0.00    0.00
Combined Print and E-Book Nonfiction   0.000000   0.00    0.00
Crime and Punishment                   0.000000   0.00    0.00
Culture                                0.000000   0.00 

In [67]:
# 12. Group the dataset by two columns and then sort the aggregated results within the groups
grouped_by_book_material_author = df_cleaned_sorted.groupby(['Book_Material', 'author']).agg({'rank': 'mean'}).sort_values(by='rank', ascending=True)

# Displaying the results
print("\nGrouped by book material and author (sorted by mean rating):")
print(grouped_by_book_material_author)


Grouped by book material and author (sorted by mean rating):
                                                                           rank
Book_Material                        author                                    
Hardcover Graphic Books              J Michael Straczynski and Don Kramer   1.0
Combined Print and E-Book Fiction    Robert Jordan and Brandon Sanderson    1.0
E-Book Fiction                       Tom Clancy with Mark Greaney           1.0
Combined Print and E-Book Fiction    S C Stephens                           1.0
Religion Spirituality and Faith      Marianne Williamson                    1.0
...                                                                         ...
Manga                                Tomu Ohmi                              5.0
Combined Print and E-Book Nonfiction Richard D Mahoney                      5.0
Paperback Nonfiction                 Robert M Sapolsky                      5.0
E-Book Nonfiction                    Michael Hainey       

# Conclusions  

After exploring your dataset, provide a short summary of what you noticed from this dataset.  What would you explore further with more time?

After exploring the New York Times Best Sellers dataset, I noticed that the average price for purchasing Manga was $5.16. The authors: J Michael Straczynski and Don Kramer for Hardcover Graphic Books was ranked #1. To add on, there were 233 books that Stephen King wrote. In the future, I would like to explore other types of books authored by Japanese Authors by filtering the dataset.